### ISSUE TITLE
Warning when passing `optimizer` through `mmrm` to `h_get_optimizers`

### ISSUE THREAD

**Issue by dgkf** [openpharma/mmrm#221]
**Summary**

When passing an `optimizer` (singular) to `mmrm` you get the warning:
```
Warning: partial match of `optimizer` to `optimizers`
```

This is emitted when running our test suit, which gives the `testthat` warning:
```
Warning (test-fit.R:304): mmrm falls back to other optimizers if default does not work
partial match of 'optimizer' to 'optimizers'
```

`mmrm_control` accepts argument `optimizers` which expects a `h_get_optimizers` object.
`h_get_optimizers` accepts argument `optimizer` which expects a character vector of optimizer names.

However, `mmrm_control` is intercepting argument `optimizer` as it is partially matched with its own argument `optimizers`, meaning that the optimizers are taking the wrong data type.

**Comment by danielinteractive** [comment]
@clarkliming maybe you would like to have a look into it?

**Comment by clarkliming** [comment]
@dgkf could you please provide more details on this warning? i.e. platform, code executed, package versions, etc.

I am asking this because I did not run into this issue before, and the checks on github seems to work well.

Here is part of the code, where `optimizers` is after `...`. In this case, the partial matches should not intercept the partial matches because they should go into the `...`
```
mmrm_control <- function(n_cores = 1L,
                         method = c("Satterthwaite", "Kenward-Roger", "Kenward-Roger-Linear"),
                         start = NULL,
                         accept_singular = TRUE,
                         drop_visit_levels = TRUE,
                         ...,
                         optimizers = h_get_optimizers(...)) {
```

**Comment by dgkf** [comment]
Ah, I found it's because I have `options(warnPartialMatchDollar = TRUE)`, which is identifying this misnamed dollar access:

https://github.com/openpharma/mmrm/blob/11e3b2d0f28cf98a526843f240f8ae8a6c79a92d/R/fit.R#L356-L360

Which should be

```diff
-        "Chosen optimizer '", toString(names(control$optimizer)), "' led to problems during model fit:\n",
+        "Chosen optimizer '", toString(names(control$optimizers)), "' led to problems during model fit:\n",
```

---

Investigating a bit _why_ this isn't breaking things, I was surprised to peel back R's gnarly shell and discover a new R idiosyncrasy!

I was already aware that R "helpfully" matches partial argument names:

```r
f <- function(aaaaaaaaa = "default", ...) cat(aaaaaaaaa, "\n")
f(a = "not default!")
# not default!
```

But apparently this only happens for arguments _before_ the ellipsis, so this will _not_ match:

```r
f <- function(..., aaaaaaaaa = "default") cat(aaaaaaaaa, "\n")
f(a = "not default!")
# default
```

In our case, `mmrm_control`'s formal arguments are dependent on `optimizers` coming after `...` in `mmrm_control` so that `optimizer` (intended for `h_get_optimizers`) isn't intercepted by `mmrm_control`'s `optimizers` parameter. Maybe this was part of the design, but I would consider this a precarious behavior to rely on as it's quite intolerant to minor refactoring.

**Comment by clarkliming** [comment]
> optimizers

maybe we should consider renaming the argument `optimizer` and `optimizers`; it is confusing because optimizers are the optimizer functions we used while `optimizer` are just string representations of commonly used ones. 

Personally I don't like this feature of partial matching in R, but it seems that what we could do is to make the arguments very different from each other...

**Comment by danielinteractive** [comment]
I would vote for consider renaming the arguments to be more different. And agree that we should make the functions that user could use part of the API

**Comment by dgkf** [comment]
At least as far as these warnings are concerned, they were resolved in #214. I'm going to close this and we can file a separate issue to discuss argument conventions.
